### PR TITLE
[gnocchi] New Gnocchi plugin

### PR DESCRIPTION
**Pull request by psachin**
Following files are collected:
- /etc/gnocchi/gnocchi.conf
- /etc/gnocchi/gnocchi-config-generator.conf
- /etc/gnocchi/policy.json
- /etc/gnocchi/api-paste.ini

Following outputs are collected:
- `gnocchi --version`
- `gnocchi status`
- `gnocchi capabilities list`
- `gnocchi archive-policy list`
- `gnocchi resource list`
- `gnocchi resource-type list`

Grafana commands ouputs are captured('-o grafana -k gnocchi.grafana=True'):
- `grafana-cli --version`
- `grafana-cli plugins ls`

Closes-Bug: #678

Signed-off-by: Sachin Patil psachin@redhat.com

### CODE REVIEW

File: sos/plugins/gnocchi.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    def setup(self):
+        self.add_cmd_output([
+            "gnocchi --version",
+            "gnocchi status",
+            "gnocchi capabilities list",
+            "gnocchi archive-policy list",
+            "gnocchi resource list",
+            "gnocchi resource-type list"
+        ])
```
Comment (bmr-cymru) [inline sos/plugins/gnocchi.py]: I see that the `gnocchi` command accepts authentication information via the `--os-username`, `--os-password`, and related environment variables (`OS_USER` etc.).

Do we expect that all installations will have this disabled? If not, and the `gnocchi` shell command prompts for them interactively then `sos` will take a timeout (default 300s, 5m) on each `gnocchi` invocation.

If we do expect to find authenticated installations then the plugin needs to be able to handle that: e.g. by passing in values in environment variables (*not* using plugin options - these are exposed in `/proc` to unprivileged users).

**Comment by bmr-cymru** [comment]
Ready for merge once the authentication questions are sorted out.

**Comment by psachin** [comment]
I've see other plugins like openstack_nova also using command when proper environment variable are set. I decided to go same way and tested on packstack setup by sourcing `~/keystone_admin` file.

`cmds` will be used as plugin option.

I see many contributors using 
```
        if self.get_option('cmds'):
            for os_var in ['OS_USERNAME', 'OS_PASSWORD', 'OS_TENANT_NAME']:
                if os_var not in os.environ:
                    self.soslog.warning("%s not found in environment variables"
                                        " which is required" % (os_var))

            self.add_cmd_output([...
```
Can this be refactored under `class Plugin` like we have `option_list`? We can have something like
`check_os_env['OS_USERNAME', 'OS_PASSWORD'..]`. Just a thought.

Also as looked as @cshastri working on #903 on which you suggested to use `do_file_sub()` for single file. I liked the suggestion and followed the same again.

**Comment by bmr-cymru** [comment]
> I see many contributors using

I would not say 'many' - the only plugin that does this is `openstack_nova`, and I would consider it buggy. There's no reason to disable all command collection for a plugin - if particular commands are slow/expensive we already have mechanisms for dealing with that.

So the question is: why does this plugin need this special 'cmds' option magic, when nothing else does?

It's much better from a support data collection, and user experience standpoint, to always attempt to collect these commands when the relevant environment variables are set. Gating it on an additional per-plugin option just makes things hard work for everyone.

It's fine to emit a log message if those variables are *not* set, so that a support associate viewing the data knows what's going on, but relying on end-users to add (multiple) plugin options is just asking for missing data.

**Comment by psachin** [comment]
@bmr-cymru 

Ya..that makes sense. It should reduce to-and-fro conversation with customers(asking them to run sosreport again with options). Anyway's I have added verbose log message which should be helpful for new associates to justify back to end-users in case we again ask for logs

**Comment by bmr-cymru** [comment]
Thanks! I'll get these included in the next round of thorough merge reviews.

**Comment by bmr-cymru** [comment]
I don't think GitHub understand the "Closes-Bug" notation you used in your commit message - either "Closes:" or "Resolves:" work (there's a list somewhere in the GH help).

I'll mark this closed in commit 1d66761d6bd95e205f4b2495e9ae40681958c552 manually.

**Pull request closed without merging**